rfd/0181-pg-database-access-webui.md

```diff
@@ -0,0 +1,322 @@
+will generate a user certificate containing `RouteToDatabase` using one of
+`PerformMFACeremony` or `GenerateUserCerts`.
+
+Note: The MFA flows will follow the existing functions (used by Kubernetes
+execs, desktops, and SSH sessions).
+
+Once the certificates are generated, the proxy will connect to the target
+database server. This flow resembles what the database proxy does: Choose a
+database server among a list of available servers to forward the connection.
+After the connection is established, a new instances of the Teleport's
+PostgreSQL REPL is created, and the session starts.
+
+#### PostgreSQL REPL
+
+The proxy will include a PostgreSQL Read-Eval-Print Loop (REPL) that interacts
+with the databases served by database servers. The REPL will read the user input
+per line and decide whether or not to execute the command (as the REPL will
+support multi-line commands). The rule follows:
+- If the line starts with a backslash command, execute it.
+- Send the line to the target database if it terminates in a semicolon (`;`).
+
+If the line read doesn't meet the above criteria, it is added to an internal
+buffer and is accumulated until it is ready to be executed.
+
+The REPL will interact with the WebSocket connection using the existent
+`terminal` (`lib/web/terminal/terminal.go`) structs.
+
+##### Backslash commands
+
+Backslash commands are executed at the proxy, and results are forwarded to the
+client's websocket. Depending on the command executed, there is no interaction
+with the database server, and the command won't be included in the session
+recording.
+
+There are two categories of backslash commands: those that don't require
+database interaction (for example, printing the help menu) and those that expand
+into SQL queries forwarded into the database.
+
+For the latter, we'll keep a list of pre-written queries based on the queries
+used at `psql`.
+
+##### Command/Queries execution
+
+To execute commands and queries into the target PostgreSQL, the REPL will use a
+`pgx.Conn` instead of manually generating PostgreSQL protocol messages. This
+will simplify the protocol interaction and cover most use cases.
+
+The `pgx.Conn` will be started during the REPL initialization process.
+
+##### Result formatting
+
+Once the query is executed, the interactive shell will parse command tag results
+and decide how to render it:
+- For `SELECT` commands: Generate an ASCII table using the query results with
+  the same format as `psql`. The field information (table head) will be fetched
+  using the result field descriptors.
+- For other commands: Display the command tag as it is received.
+
+```shell
+postgres=# SELECT * FROM events;
+
+id | name
+---+-------
+ 1 | start 
+ 2 | end 
+(2 rows)
+
+postgres=# INSERT INTO events (name) VALUES ('end');
+INSERT 0 1
+```
+
+Errors will be shown with `ERROR:` prefix.
+
+```shell
+postgres=> SELECT err;
+ERROR: column "err" does not exist
+
+```
+
+### Security
```

Comment:
could we add a section on product metrics? like how would we measure if ppl are using this feature and if it's successful.